lld/ELF/Driver.cpp

```diff
@@ -2619,6 +2623,42 @@ static uint32_t getAndFeatures() {
   return ret;
 }
 
+static void getAArch64PauthInfo() {
```

Comment:
I think `uint32_t getAndFeatures` can be renamed to `void readSecurityNotes`. `readSecurityNotes` sets `andFeatures` internally, and  `getAArch64PauthInfo` is merged into it.

This way we iterate objectFiles just once for aarch64.

Reply:
I've renamed `uint32_t getAndFeatures` to `void readSecurityNotes` and merged `getAArch64PauthInfo` into it. As far as I understood, your intention was to avoid multiple loops over `ctx.objectFiles`, so, I've put processing of both `andFeatures` and `aarch64PauthAbiTag` into a single loop. I'm not sure if it increases readability though - I personally prefer smaller functions doing simple separate tasks. And we also still need `std::find_if` over `ctx.objectFiles` before the loop to find the first non-empty tag in order to emit `-z pauth-report`-related messages. Such `std::find_if` is effectively an extra iteration over `ctx.objectFiles` which can't be easily avoided, and I'm not sure if handling both `andFeatures` and `aarch64PauthAbiTag` in a single loop inside a single function gives any extra value.

See 1448a4a4e0e3cb70358558579e97fce3f0990356